### DISCUSSION TITLE
Is it practical to implement a type that supports arbitrary (more or less) dot access?

### DISCUSSION THREAD

**Discussion by jimfulton** [sqlalchemy/sqlalchemy#6934] [Usage Questions]
Hi,

For BigQuery, we currently represent [struct](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#struct_type) (aka record) values as JSON.  To access fields, we'll have to use mapping syntax with string field names.  This isn't horrible. :)

I'm wondering if I should try to make dot syntax (`mytable.c.mystruct.myfield`) possible, but I don't see a way to do it, and can think of reasons why it would be problematic. :)

Am I missing something? Is there a reasonable way to make dot access work for most names (excluding some set of special names)?

**Comment by zzzeek** [comment]
Hmmm well mapping syntax which I assume you mean `foo['bar']` has the advantage that it falls under Python's operator overrides, in this case `__getitem__()`, unambiguously.  For dot access, we of course use `__getattr__()`, and I'm sure you are aware of the pit of despair this can lead towards, as it then becomes a guessing game when someone wants to just access an actual member of the object that can't be found, etc.

Within SQLAlchemy's operator system, the base of which you can take a look at [here](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/sql/operators.py), you can see which Python special methods we add operator semantics towards, and `__getattr__()` is obviously not one of them since it's pretty special.

Finally there's the simple fact of naming collisions, `mytable.c.mystruct` is a `Column` object and it has lots of attributes of its own that mean something, using such unusual names as `.name`, `.key`, `.info`, `.default`, `.system` and other names that I'm sure nobody would ever have in their structs :)    this is why `Table` has that `.c` namespace, so that there are no naming collisions.     

What this *does* indicate is that perhaps `Column` *could* have it's own version of `.c` for this purpose, like `.attr` / `.a` for shorthand (note that `.c` is short for `.columns` which is always there too).    So not today, but someday you could consider a syntax like `mytable.c.mystruct.a.myfield` as more feasible perhaps.   just an idea as i was typing this.

**Reply by jimfulton** [reply]
> What this _does_ indicate is that perhaps `Column` _could_ have it's own version of `.c` for this purpose, like `.attr` / `.a` for shorthand (note that `.c` is short for `.columns` which is always there too). So not today, but someday you could consider a syntax like `mytable.c.mystruct.a.myfield` as more feasible perhaps. just an idea as i was typing this.

Yeah, worth pondering.
